theforeman.org/pipelines/lib/packaging.groovy

```diff
@@ -0,0 +1,455 @@
+        echo "echo deb http://deb.theforeman.org/ ${os} ${version} >> /etc/apt/sources.list" | sudo tee -a \$addrepo_hook > /dev/null
+        echo "echo deb http://deb.theforeman.org/ plugins ${version} >> /etc/apt/sources.list" | sudo tee -a \$addrepo_hook > /dev/null
+
+        # Read Foreman package installation logs on failure
+        echo "cat /var/log/foreman-install.log || true" | sudo tee \${pbuilder_hooksdir}/C10foremanlog
+
+        echo "tail -n 100 /etc/apt/sources.list /etc/apt/sources.list.d/*.list" | sudo tee \${pbuilder_hooksdir}/F99printrepos
+
+        # Make executable
+        sudo chmod 0775 \${pbuilder_hooksdir}/*
+    """
+}
+
+def deb_build_dir(type, project, os = null) {
+    def build_dir
+
+    if (type == 'core') {
+        build_dir = "debian/${os}/build-${project}"
+    } else if (type == 'dependency') {
+        build_dir = "dependencies/${os}/build-${project}"
+    } else if (type == 'plugin') {
+        build_dir = "plugins/build-${project}"
+    } else {
+        echo "Unsupported type specified: ${type}"
+        exit 1
+    }
+
+    return build_dir
+}
+
+def setup_sources_deb(type, project, os, version, repoowner, pull_request) {
+
+    if (type == 'core') {
+        setup_sources_core(project, os, version, repoowner, pull_request)
+    } else if (type == 'dependency') {
+        setup_sources_dependency(project, os, version, repoowner, pull_request)
+    } else if (type == 'plugin') {
+        setup_sources_plugin(project, os, version, repoowner, pull_request)
+    } else {
+        echo "Unsupported type specified: ${type}"
+        exit 1
+    }
+
+}
+
+def setup_sources_core(os, project, version, repoowner, pull_request = false) {
+    def package_version = sh(returnStdout: true, script: "head -n1 debian/${os}/${project}/changelog|awk '{print \$2}'|sed 's/(//;s/)//'|cut -f1 -d-").trim()
+    def build_dir = deb_build_dir('core', project, os)
+
+    package_version = package_version.replaceAll('~rc', '-rc')
+    last_commit = package_version
+
+    sh "mkdir -p ${build_dir}"
+
+    dir(build_dir) {
+        if (version =~ 'develop') {
+            def url_base = 'https://ci.theforeman.org'
+            def job_name = nightly_jenkins_job
+            def job_id = nightly_jenkins_job_id
+            def json_url = "${url_base}/job/${job_name}/${job_id}/api/json"
+
+            sh """
+              # If a last* alias was used, resolve the numeric job ID
+              job_id=\$(curl "${json_url}" | /usr/local/bin/JSON.sh -b | awk '\$1 == "[\"number\"]" { print \$2 }')
+              base_url=`curl "${json_url}" | /usr/local/bin/JSON.sh -b | awk '\$1 ~ /^\\["runs",.*,"number"\\]/ && \$2 == '\$job_id' {getline; print \$2; exit}' | tr -d \"`
+
+              if [ x\$base_url = x ] ; then
+                base_url=`curl "${json_url}" | /usr/local/bin/JSON.sh -b | egrep '\\["url"\\]' | awk '{print \$NF}' | tr -d \"`
+              fi
+              url="${base_url}/artifact/*zip*/archive.zip"
```

Comment:
This can be simplified if we merge https://github.com/theforeman/jenkins-jobs/pull/90 first :)